### PR TITLE
MRG, FIX: Allow pickling

### LINKED ISSUE #602: Building fails with pickling error
Hi, I'm trying to build a gallery for a project that uses the multiprocessing module (hence, it needs to use the pickle module), the script defines a few functions that have to be pickled but building the gallery throws the following error (_get is the function that fails to pickle):

```_pickle.PicklingError: Can't pickle <function _get at 0x11cc9c0d0>: attribute lookup _get on __main__ failed```

Full traceback:

```
Exception occurred:
  File "/Users/Edu/miniconda3/envs/ploomber/lib/python3.6/site-packages/sphinx_gallery/gen_gallery.py", line 544, in summarize_failing_examples
    "\n" + "-" * 79)
ValueError: Here is a summary of the problems encountered when running the examples

Unexpected failing examples:
/Users/Edu/dev/ploomber/examples/train-and-predict.py failed leaving traceback:
Traceback (most recent call last):
  File "/Users/Edu/miniconda3/envs/ploomber/lib/python3.6/site-packages/sphinx_gallery/gen_rst.py", line 440, in _memory_usage
    out = func()
  File "/Users/Edu/miniconda3/envs/ploomber/lib/python3.6/site-packages/sphinx_gallery/gen_rst.py", line 425, in __call__
    exec(self.code, self.globals)
  File "/Users/Edu/dev/ploomber/examples/train-and-predict.py", line 124, in <module>
    dag_fit.build()
  File "/Users/Edu/dev/ploomber/src/ploomber/dag.py", line 167, in build
    return self._executor(dag=self, force=force)
  File "/Users/Edu/dev/ploomber/src/ploomber/executors/Serial.py", line 45, in __call__
    raise e
  File "/Users/Edu/dev/ploomber/src/ploomber/executors/Serial.py", line 40, in __call__
    t.build(**kwargs)
  File "/Users/Edu/dev/ploomber/src/ploomber/tasks/Task.py", line 303, in build
    raise e
  File "/Users/Edu/dev/ploomber/src/ploomber/tasks/Task.py", line 293, in build
    self.run()
  File "/Users/Edu/dev/ploomber/src/ploomber/tasks/tasks.py", line 86, in run
    res.get()
  File "/Users/Edu/miniconda3/envs/ploomber/lib/python3.6/multiprocessing/pool.py", line 644, in get
    raise self._value
  File "/Users/Edu/miniconda3/envs/ploomber/lib/python3.6/multiprocessing/pool.py", line 424, in _handle_tasks
    put(task)
  File "/Users/Edu/miniconda3/envs/ploomber/lib/python3.6/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/Users/Edu/miniconda3/envs/ploomber/lib/python3.6/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <function _get at 0x11cc9c0d0>: attribute lookup _get on __main__ failed
```

The error happens in ```sphinx_gallery/gen_rst.py```, looks like the way sphinx-gallery runs the code makes the function declaration unavailable, that's why pickle fails.

I noticed that if I define the function in a separate file and import it, the build succeeds. But then the example is incomplete since the source code for the function does not appear.

As a workaround: is there any way for sphinx-gallery to build examples so that the source code for both files is shown?

### PR DESCRIPTION
**Pull request by larsoner**
Closes #602.

Works around our use of `exec` by doing a better job of faking that things are being executed in `__main__`.

This isn't really mentioned in the [links](https://medium.com/@jwnx/multiprocessing-serialization-in-python-with-pickle-9844f6fa1812) that I [looked at](https://stackoverflow.com/questions/43794595/pickle-doesnt-work-in-compile-exec) to find [ideas](https://stackoverflow.com/questions/31191947/pickle-and-exec-in-python). I wonder if there is a good reason if this wasn't suggested (i.e., it's a bad idea) or if it wasn't considered.

- [x] make CIs happy
- [x] get confirmation that it fixes #602
- [x] ~~Add a `multiprocessing` example~~
- [x] Add a `joblib` example
- [x] locally build MNE-Python doc while using `joblib` to see if it still works: https://github.com/mne-tools/mne-python/pull/7274 / https://17777-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_forward.html#compute-forward-solution

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
Very nice!!

This first iteration looks good to me. I'm waiting for you to tick the rest of the boxes on the TODO list.

**Comment by larsoner** [comment]
Okay boxes ticked. Didn't do `multiprocessing` explicitly because I think `joblib` is enough.

**Comment by larsoner** [comment]
@lucyleeow or @choldgraf this should be good to go

**Comment by larsoner** [comment]
@lucyleeow this one should be good to go, too

File: sphinx_gallery/gen_rst.py [hunk 1]
```diff
@@ -415,16 +417,22 @@ def patch_warnings():
...
             self.run = True
+            old_main = sys.modules.get('__main__', None)
             with patch_warnings():
-                exec(self.code, self.globals)
+                sys.modules['__main__'] = self.fake_main
+                try:
+                        exec(self.code, self.fake_main.__dict__)
+                finally:
+                    if old_main is not None:
+                        sys.modules['__main__'] = old_main
```
Comment (lucyleeow) [inline sphinx_gallery/gen_rst.py]: What is the purpose of using just `try` and `finally`?

File: sphinx_gallery/gen_rst.py [hunk 2]
```diff
@@ -626,20 +642,23 @@ def execute_script(script_blocks, script_vars, gallery_conf):
     time_elapsed : float
         Time elapsed during execution
     """
-
-    example_globals = {
+    # Examples may contain if __name__ == '__main__' guards
+    # for in example scikit-learn if the example uses multiprocessing.
+    # Here we create a new __main__ module, and temporarily change
+    # sys.modules when running our example
+    fake_main = imp.new_module("__main__")
```
Comment (lucyleeow) [inline sphinx_gallery/gen_rst.py]: Should we use `importlib.util.module_from_spec()` instead as `imp` is deprecated? https://docs.python.org/3/library/imp.html#imp.new_module

**Comment by lucyleeow** [comment]
Thanks for the pickle links as well - they were very useful.

File: sphinx_gallery/gen_rst.py (hunk 1 above)
Comment (larsoner) [reply sphinx_gallery/gen_rst.py]: It is equivalent to:
```
try:
    ...
except:
    raise
finally:
    ...
```
i.e., it always does what is in the `finally` clause regardless of the success or failure of the `try`. In this case we always want to restore the `sys.modules['__main__']` regardless of whether or not our `exec` worked.

Comment (lucyleeow) [reply sphinx_gallery/gen_rst.py]: Ah yes of course, thanks!

**Comment by larsoner** [comment]
Uses `importlib` now and made the `RandomState` test more robust, ready for review/merge from my end @lucyleeow

File: sphinx_gallery/tests/test_gen_rst.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
                         unicode_literals)
 import ast
 import codecs
+import imp
```
Comment (lucyleeow) [inline sphinx_gallery/tests/test_gen_rst.py]:
```suggestion
import importlib
```

File: sphinx_gallery/tests/test_gen_rst.py [hunk 4]
```diff
@@ -547,7 +548,21 @@ def test_rst_example(gallery_conf):
     assert "rst-class:: sphx-glr-timing" in rst
 
 
-def test_output_indentation(gallery_conf, tmpdir):
+@pytest.fixture(scope='function')
+def script_vars(tmpdir):
+    fake_main = imp.new_module('__main__')
```
Comment (lucyleeow) [inline sphinx_gallery/tests/test_gen_rst.py]:
```suggestion
    fake_main = importlib.util.module_from_spec(
        importlib.util.spec_from_loader('__main__'))
```

Comment (lucyleeow) [reply sphinx_gallery/tests/test_gen_rst.py]: Not sure if `None` is needed in `importlib.util.spec_from_loader('__main__')`

`importlib` is much more verbose!

**Comment by lucyleeow** [comment]
I guess you need `None`:
```
    fake_main = importlib.util.module_from_spec(
        importlib.util.spec_from_loader('__main__', None))
```

**Comment by lucyleeow** [comment]
Thanks @larsoner!

**Pull request merged**